File: AUGUST-2025/04-08-2025/p904.ex
defmodule Solution do
  @spec total_fruit(fruits :: [integer]) :: integer
  def total_fruit(fruits) do
    Enum.reduce(fruits, {0, nil, nil, 0, 0}, fn
      f1, {ans, f1, f2, len, len2} ->
        {max(ans, len + 1), f1, f2, len + 1, len2 + 1}
      f2, {ans, f1, f2, len, len2} ->
        {max(ans, len + 1), f2, f1, len + 1, 1}
      x, {ans, f1, _, _, len2} ->
        {max(ans, len2 + 1), x, f1, len2 + 1, 1}
    end)
    |> elem(0)
  end
end

File: AUGUST-2025/04-08-2025/p904.rs
impl Solution {
    pub fn total_fruit(f: Vec<i32>) -> i32 {
        let (mut p, mut k, mut j) = (-1, 0, 0);
        f.iter().enumerate().map(|(i, &t)| {
            if t != f[j] {
                if t != p { k = j }
                p = f[j]; j = i
            }
            i - k + 1
        }).max().unwrap() as _
    }

}